### PR TITLE
Remove some unused type ignores (round 1)

### PR DESCRIPTION
**Pull request by bluenote10**
Over time, a large number of the existing type ignores have become irrelevant/unused/dead as a result of improvements in annotations and type checking.

Having these `# type: ignore` linger around is not ideal for two reasons:

- They are verbose/ugly syntatically.
- They could hide genuine bugs in the future, if a refactoring would actually introduce a bug but it gets hidden by the ignore.

I'm counting over 1500 unused ignores already. This is a first PR that removes some of them. Note that I haven't touched type ignores that looked "conditional" like the import challenge mentioned in https://github.com/pytorch/pytorch/pull/60006#issuecomment-2480604728. I will address these at a later point, and eventually would enable `warn_unused_ignores = True` in the mypy configuration as discussed in that comment to prevent accumulating more dead ignores going forward.

This PR should have no effect on runtime at all.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @ezyang @SherlockNoMad

### CODE REVIEW

File: torch/ao/quantization/observer.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ class ObserverBase(ABC, nn.Module):
         or static quantization
     """
 
-    def __init__(self, dtype, is_dynamic=False):
+    def __init__(self, dtype, is_dynamic: bool = False):
```
Comment (janeyx99) [inline torch/ao/quantization/observer.py:149]: Curious—are you running another lint pass on these aside from removing unused type ignores?

**Review by janeyx99 — COMMENTED** [review]
Went through all instances in PR and they look okay. Question—did you generate these automatically with a script?

Approving workflows to ensure lint passes before approving the PR

**Comment by bluenote10** [comment]
> Went through all instances in PR and they look okay. Question—did you generate these automatically with a script?

To detect them I simply changed `warn_unused_ignores` in the `mypy.ini` from `False` to `True` temporarily. Then I had the full list unused ignores, exactly 1543 errors. I did not resolve them automatically, because I want to be careful with these "conditionally" unused ignores as mentioned in https://github.com/pytorch/pytorch/pull/60006#issuecomment-2480604728. It will take a few rounds of PRs like this one to bring the number to 0.

File: torch/ao/quantization/observer.py (hunk 1 above)
Comment (bluenote10) [reply torch/ao/quantization/observer.py:149]: In this case, that implicit `Any` just caught my eye, and I figured I add some actual type information while I'm at it.

File: torch/ao/nn/quantized/modules/linear.py [hunk 2]
```diff
@@ -309,10 +309,14 @@ def from_float(cls, mod, use_precomputed_fake_quant=False):
...
-            supported_modules = ", ".join([float_mod.__name__ for float_mod in cls._FLOAT_MODULE])  # type: ignore[attr-defined]
+                cls._FLOAT_MODULE = [cls._FLOAT_MODULE]
+            supported_modules = ", ".join(
+                [float_mod.__name__ for float_mod in cls._FLOAT_MODULE]
+            )
             error_msg = f"nnq.{cls.__name__}.from_float only works for {supported_modules}, but got: {type(mod)}"
-            assert type_before_parametrizations(mod) in cls._FLOAT_MODULE, error_msg.format()  # type: ignore[attr-defined]
+            assert (
+                type_before_parametrizations(mod) in cls._FLOAT_MODULE
+            ), error_msg.format()
```
Comment (Skylion007) [inline torch/ao/nn/quantized/modules/linear.py:319]: Not, should error_msg be wrapped in our LazyStr so the formatting only happens when the assert triggers?

File: torch/jit/_serialization.py [hunk 3]
```diff
@@ -149,10 +149,10 @@ def load(f, map_location=None, _extra_files=None, _restore_shapes=False):
         os.remove("scriptmodule.pt")
     """
     if isinstance(f, (str, os.PathLike)):
```
Comment (Skylion007) [inline torch/jit/_serialization.py:151]: I think was also originally supposed to accept bytes, and might have been killed in some refactor. Hence the type:ignore

**Review by Skylion007 — COMMENTED** [review]
Some nits/comments.

File: torch/ao/nn/quantized/modules/linear.py (hunk 2 above)
Comment (bluenote10) [reply torch/ao/nn/quantized/modules/linear.py:319]: That sounds sensible, but I may be lacking context to judge this.

In any case this change should probably not be made as part of this PR right? This PR is just about remove type ignores that aren't needed, trying not to mix it with any behavioral changes.

File: torch/jit/_serialization.py (hunk 3 above)
Comment (bluenote10) [reply torch/jit/_serialization.py:151]: Yes, that would also have been my guess. In any case, there is currently no type error here.

**Review by Skylion007 — APPROVED** [review]

**Comment by ezyang** [comment]
I haven't reviewed, but thanks! This stuff is much appreciated

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**